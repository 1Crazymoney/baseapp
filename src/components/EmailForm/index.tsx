import {
    Button,
    Input,
    Loader,
} from '@openware/components';
import cr from 'classnames';
import * as React from 'react';
import {
    EMAIL_REGEX,
} from '../../helpers';

interface EmailFormProps {
    title?: string;
    buttonLabel?: string;
    errorMessage?: string;
    isLoading?: boolean;
    OnSubmit: () => void;
    className?: string;
    emailLabel?: string;
    email: string;
    emailError: string;
    emailFocused: boolean;
    validateForm: () => void;
    handleInputEmail: (value: string) => void;
    handleFieldFocus: () => void;
}

class EmailForm extends React.Component<EmailFormProps> {
    public render() {
        const {
            title,
            buttonLabel,
            isLoading,
            emailLabel,
            email,
            emailFocused,
            emailError,
        } = this.props;
        const emailGroupClass = cr('cr-email-form__group', {
            'cr-email-form__group--focused': emailFocused,
        });
        return (
            <form>
                <div className="cr-email-form">
                    <div className="cr-email-form__options-group">
                        <div className="cr-email-form__option">
                            <div className="cr-email-form__option-inner">
                                {title ? title : 'Forgot password'}
                            </div>
                        </div>
                    </div>
                    <div className="cr-email-form__form-content">
                        <div className={emailGroupClass}>
                            <label className="cr-email-form__label">
                                {emailLabel ? emailLabel : 'Email'}
                            </label>
                            <Input
                                type="email"
                                value={email}
                                className="cr-email-form__input"
                                onChangeValue={this.props.handleInputEmail}
                                onFocus={this.props.handleFieldFocus}
                                onBlur={this.props.handleFieldFocus}
                            />
                            {emailError && <div className="cr-email-form__error">{emailError}</div>}
                        </div>
                        <div className="cr-email-form__button-wrapper">
                            <div className="cr-email-form__loader">{isLoading ? <Loader/> : null}</div>
                            <Button
                                label={isLoading ? 'Loading...' : buttonLabel ? buttonLabel : 'Send'}
                                type="submit"
                                className="cr-email-form__button"
                                disabled={isLoading}
                                onClick={this.handleClick}
                            />
                        </div>
                    </div>
                </div>
            </form>
        );
    }

    private handleSubmitForm() {
        this.props.OnSubmit();
    }

    private isValidForm() {
        const { email } = this.props;
        const isEmailValid = email.match(EMAIL_REGEX);

        return email && isEmailValid;
    }

    private handleClick = (label?: string, e?: React.FormEvent<HTMLInputElement>) => {
        if (e) {
            e.preventDefault();
        }
        if (!this.isValidForm()) {
            this.props.validateForm();
        } else {
            this.handleSubmitForm();
        }
    };
}


export {
    EmailForm,
    EmailFormProps,
};
