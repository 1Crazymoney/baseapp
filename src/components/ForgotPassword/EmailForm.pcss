.cr-email-form {
  background: var(--email-form-bg);
  max-width: 424px;
  min-height: 350px;
  box-shadow: 0 10px 15px rgba(0, 0, 0, 0.25);
  border-radius: 3px;
  position: relative;

  &__options-group {
    width: 100%;
    height: 58px;
    border-bottom: 2px solid var(--email-form-border);
  }

  &__option {
    background: var(--header);
    display: inline-block;
    width: 100%;
    height: 58px;
  }

  &__option-inner {
    background: var(--transparent);
    width: 100%;
    height: 58px;
    margin: 0 auto;
    text-align: center;
    text-transform: uppercase;
    line-height: 58px;
    font-style: normal;
    font-weight: 500;
    font-size: 14px;
    letter-spacing: 0.0357143em;
    color: var(--color-white);
  }

  &__form-content {
    position: absolute;
    top: 80px;
    right: 23px;
    bottom: 23px;
    left: 23px;
    box-sizing: border-box;
  }

  & > * {
    background: var(--transparent);
  }

  &__error-message {
    color: var(--color-red);
    font-size: calc(var(--gap) * 2.5);
    flex: 1;
    display: flex;
    background: var(--transparent);
    justify-content: center;
    align-items: center;
    min-height: calc(var(--gap) * 6);
  }

  &__group {
    background: var(--transparent);
    margin-bottom: 20px;
    position: relative;
  }

  &__error {
    background: var(--transparent);
    margin-top: 5px;
    color: var(--color-red);
  }

  &__label {
    position: relative;
    top: 16px;
    left: 10px;
    padding: 2px 6px;
    border-radius: 2px;
    background: var(--email-form-bg);
    color: var(--forgot-password-label-color);
  }

  &__group--focused {
    .cr-email-form__label {
      color: var(--email-form-accent);
    }

    input {
      border: 1px solid var(--email-form-accent) !important;
    }
  }

  &__image {
    display: block;
    width: 50%;
    height: 100px;
    margin: 0 auto 20px auto;
  }

  &__title {
    color: var(--forgot-password-label-color);
    margin: calc(var(--gap) * 3) 0 calc(var(--gap) * 8) 0;
    font-size: 18px;
    text-align: center;
  }

  &__button {
    margin-left: 0;
    height: 40px;
    background: var(--button-primary);
    border-radius: 3px;
    font-style: normal;
    font-weight: normal;
    line-height: normal;
    font-size: 16px;
    text-align: center;
    text-transform: uppercase;
    color: var(--color-white);
  }

  &__loader {
    background: var(--transparent);
    min-height: calc(var(--gap) * 6);
  }

  &__button-wrapper {
    background: var(--transparent);
    min-height: calc(var(--gap) * 28);
    display: flex;
    flex-direction: column;
    justify-content: space-between;
    position: absolute;
    right: 0;
    bottom: 0 !important;
    left: 0;
  }

  &__button-wrapper--empty {
    justify-content: flex-end;
  }

  &__input {
    width: 100%;
    margin-top: 8px;

    input {
      color: var(--forgot-password-input-color);
      width: 100%;
      height: 50px;
      padding: 0 10px;
      font-size: 12px;
      text-align: left;
      border: 1px solid var(--unknown-colors);
      box-sizing: border-box;
      border-radius: 4px;
    }
  }
}
