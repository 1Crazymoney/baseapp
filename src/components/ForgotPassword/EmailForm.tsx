/* tslint:disable:jsx-no-lambda*/
import {
    Button,
    Input,
    Loader,
} from '@openware/components';
import cr from 'classnames';
import * as React from 'react';
import {
    EMAIL_REGEX,
    ERROR_INVALID_EMAIL,
} from '../../helpers';

interface EmailFormProps {
    title?: string;
    buttonLabel?: string;
    errorMessage?: string;
    isLoading?: boolean;
    OnSubmit: (email: string) => void;
    className?: string;
    emailLabel?: string;
}

export interface EmailFormState {
    email: string;
    emailError: string;
    emailFocused: boolean;
    emailPlaceholder: string;
}

class EmailForm extends React.Component<EmailFormProps, EmailFormState> {
    constructor(props: EmailFormProps) {
        super(props);
        this.state = {
            email: '',
            emailError: '',
            emailFocused: false,
            emailPlaceholder: 'email',
        };
    }

    public render() {
        const {
            email,
            emailFocused,
            emailError,
        } = this.state;
        const {
            title,
            buttonLabel,
            errorMessage,
            isLoading,
            emailLabel,
        } = this.props;
        const buttonWrapperClass = cr('cr-email-form__button-wrapper', {
            'cr-email-form__button-wrapper--empty': !errorMessage,
        });
        const emailGroupClass = cr('cr-email-form__group', {
            'cr-email-form__group--focused': emailFocused,
        });
        return (
            <form>
                <div className="cr-email-form">
                    <div className="cr-email-form__options-group">
                        <div className="cr-email-form__option">
                            <div className="cr-email-form__option-inner">
                                {title ? title : 'Forgot password'}
                            </div>
                        </div>
                    </div>
                    <div className="cr-email-form__form-content">
                        <div className={emailGroupClass}>
                            <label className="cr-email-form__label">
                                {emailLabel ? emailLabel : 'Email'}
                            </label>
                            <Input
                                type={'mail'}
                                value={email}
                                className={'cr-email-form__input'}
                                onChangeValue={value => this.handleInput(value)}
                                onFocus={() => this.handleFieldFocus()}
                                onBlur={() => this.handleFieldFocus()}
                            />
                            {emailError && <div className={'cr-email-form__error'}>{emailError}</div>}
                        </div>
                        <div className={buttonWrapperClass}>
                            <div className="cr-email-form__error-message">{errorMessage || null}</div>
                            <div className="cr-email-form__loader">{isLoading ? <Loader/> : null}</div>
                            <Button
                                label={isLoading ? 'Loading...' : buttonLabel ? buttonLabel : 'Send'}
                                type="submit"
                                className={'cr-email-form__button'}
                                disabled={isLoading}
                                onClick={this.handleClick}
                            />
                        </div>
                    </div>
                </div>
            </form>
        );
    }

    private handleFieldFocus() {
        this.setState({
            emailFocused: !this.state.emailFocused,
        });
    }

    private handleSubmitForm() {
        const { email } = this.state;

        this.setState({
            emailError: '',
        }, () => {
            this.props.OnSubmit(email);
        });
    }

    private isValidForm() {
        const {email} = this.state;
        const isEmailValid = email.match(EMAIL_REGEX);

        return email && isEmailValid;
    }

    private validateForm() {
        const {email} = this.state;
        const isEmailValid = email.match(EMAIL_REGEX);

        if (!isEmailValid) {
            this.setState({
                emailError: ERROR_INVALID_EMAIL,
            });
            return;
        }
    }

    private handleInput = (value: string) => {
        this.setState({
            email: value,
        });
    };

    private handleClick = (label?: string, e?: React.FormEvent<HTMLInputElement>) => {
        if (e) {
            e.preventDefault();
        }
        if (!this.isValidForm()) {
            this.validateForm();
        } else {
            this.handleSubmitForm();
        }
    };
}


export {
    EmailForm,
    EmailFormProps,
};
