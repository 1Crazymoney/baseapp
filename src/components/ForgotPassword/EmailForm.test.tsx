import { shallow } from 'enzyme';
import * as React from 'react';
import { SinonSpy, spy } from 'sinon';
import { EmailForm, EmailFormProps } from './EmailForm';

const defaults: EmailFormProps = {
    OnSubmit: jest.fn(),
};

const setup = (props: Partial<EmailFormProps> = {}) =>
    shallow(<EmailForm {...{ ...defaults, ...props }} />);

describe('Email form', () => {
    let wrapper = setup();

    beforeEach(() => {
        wrapper = setup();
    });

    it('should render', () => {
        expect(wrapper).toMatchSnapshot();
    });

    it('should have correct className', () => {
        expect(wrapper.hasClass('cr-email-form'));
    });

    it('should handle Send button onClick', () => {
        const onClick = spy();
        wrapper = setup({ OnSubmit: onClick });
        wrapper.setState({ email: 'potato@heliostech.fr' });
        wrapper.find('.cr-email-form__button').simulate('click');
        expect((onClick as SinonSpy).calledOnceWith('potato@heliostech.fr')).toBeTruthy();
    });

    it('should handle input valid email value', () => {
        wrapper.props().children.props.children[1].props.children[0].props.children[1].props.onChangeValue('potato@heliostech.fr');
        // @ts-ignore
        const state: EmailFormState = wrapper.state();
        expect(state.email).toEqual('potato@heliostech.fr');
    });

    it('should handle input invalid email value', () => {
        wrapper.setState({ email: 'potato' });
        wrapper.find('.cr-email-form__button').simulate('click');
        // @ts-ignore
        const state: EmailFormState = wrapper.state();
        expect(state.emailError).toEqual('Email is invalid');
    });

    it('should render Loader', () => {
        // tslint:disable-next-line:no-magic-numbers
        expect(wrapper.props().children.props.children[1].props.children[1].props.children[1].props.children).toBeNull();
        wrapper = setup({ isLoading: true });
        // tslint:disable-next-line:no-magic-numbers
        expect(wrapper.props().children.props.children[1].props.children[1].props.children[1].props.children).not.toBeNull();
        // tslint:disable-next-line:no-magic-numbers
        expect(wrapper.props().children.props.children[1].props.children[1].props.children[2].props.label).toEqual('Loading...');
    });
});
