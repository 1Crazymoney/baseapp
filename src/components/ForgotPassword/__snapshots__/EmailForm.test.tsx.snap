// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Email form should render 1`] = `
<form>
  <div
    className="cr-email-form"
  >
    <div
      className="cr-email-form__options-group"
    >
      <div
        className="cr-email-form__option"
      >
        <div
          className="cr-email-form__option-inner"
        >
          Forgot password
        </div>
      </div>
    </div>
    <div
      className="cr-email-form__form-content"
    >
      <div
        className="cr-email-form__group"
      >
        <label
          className="cr-email-form__label"
        >
          Email
        </label>
        <e
          className="cr-email-form__input"
          onBlur={[Function]}
          onChangeValue={[Function]}
          onFocus={[Function]}
          type="mail"
          value=""
        />
      </div>
      <div
        className="cr-email-form__button-wrapper cr-email-form__button-wrapper--empty"
      >
        <div
          className="cr-email-form__error-message"
        />
        <div
          className="cr-email-form__loader"
        />
        <Component
          className="cr-email-form__button"
          label="Send"
          onClick={[Function]}
          type="submit"
        />
      </div>
    </div>
  </div>
</form>
`;
